openedx-unsupported/ecommerce
# ENT-804 Link order with enrollment code to the provided business client

@asadiqbal08 @saleem-latif @georgebabey 
**Description:** Store `Company` Field data from Basket Page in `BusinessClient` model for bulk enrollment code purchases and link the order with the that Business Client for payment processors `cybersource`, `paypal`, `stripe` if business client value is passed from basket page as `organization`.

**JIRA:** [ENT-804](https://openedx.atlassian.net/browse/ENT-804)
**Related JIRA:** [ENT-860](https://openedx.atlassian.net/browse/ENT-860)

**Implemented Tasks:**
* Add `Bulk purchase` type in the `Invoice` model and also added the related migration.
* Added a method `handle_post_order` in common order mixin .`EdxOrderPlacementMixin` which is used by all payment processor (`cybersource`, `paypal`,  `stripe`).
* Called this method `handle_post_order` immediately after an order is created to link the current order with the provided business client for bulk enrollment purchase through `Invoice` model.
* Add unittests.

**Dependencies:** https://github.com/edx/ecommerce/pull/1655

**Merge deadline:** N/A

**Installation instructions:** Checkout ecommerce on this branch `zub/ENT-804-link-enrollment-code-order-with-client` and run order model migrations like `./manage.py migrate order`.

**Testing instructions:**

1. Create a course with in **LMS** and create its related modes in **ECOM** for single seat `SKU` and bulk purchase `Bulk SKU`
<img width="1397" alt="screen shot 2018-02-19 at 5 04 37 pm" src="https://user-images.githubusercontent.com/5072991/36376985-176fab8c-1597-11e8-86f0-6d9825d48370.png">
2. In Ecom add bulk purchase product in the baseket by going to a url like: `http://localhost:8002/basket/single-item/?sku=4040F39`
<img width="883" alt="screen shot 2018-02-19 at 5 06 17 pm" src="https://user-images.githubusercontent.com/5072991/36377084-7cda0580-1597-11e8-9b0c-f94e6ba390af.png">
3. Provide the value for organization field (not present currently), e.g. `TEST CLIENT 1`. This field will be available once the related PR https://github.com/edx/ecommerce/pull/1655 is merged.
4. Click on "Place Order" button after providing the payment form values.
5. Now go to Ecom django admin for `core/businessclient` and verify that a new Business Client with the name `TEST CLIENT 1` has been added.
<img width="835" alt="screen shot 2018-02-19 at 5 11 39 pm" src="https://user-images.githubusercontent.com/5072991/36377392-dd576e56-1598-11e8-9deb-357c91ab4bbf.png">
6. Now go to Ecom django admin for `order/order` and verify that a new order has been placed with the provided payment values.
<img width="1394" alt="screen shot 2018-02-28 at 6 24 23 pm" src="https://user-images.githubusercontent.com/5072991/36789961-13539390-1cb5-11e8-8b60-3b022fa9937d.png">

7. Now go to Ecom django admin for `invoice/invoice` and verify that a new invoice has been generated against the above order.
<img width="1389" alt="screen shot 2018-02-28 at 6 24 36 pm" src="https://user-images.githubusercontent.com/5072991/36789923-ff515ee0-1cb4-11e8-9e3d-8818940da1ca.png">

8. Now verify that above invoice has type `Bulk purchase` with state `Paid` and with relation to the new business client `TEST CLIENT 1`.
<img width="706" alt="screen shot 2018-02-28 at 6 25 00 pm" src="https://user-images.githubusercontent.com/5072991/36790054-62452248-1cb5-11e8-8bfe-3300bfbfede8.png">

9. Repeat the process for single seat purchase and verify that there is no invoice record against that order (order not linked with business client).
